crates/ra_fmt2/src/edit_tree.rs

```diff
@@ -0,0 +1,115 @@
+use crate::dsl::{SpacingDsl, SpacingRule, SpaceLoc, SpaceValue};
+use crate::pattern::{Pattern, PatternSet};
+use crate::rules::spacing;
+use crate::trav_util::{walk, walk_nodes, walk_tokens};
+
+use ra_syntax::{
+    NodeOrToken, SmolStr, SyntaxElement,
+    SyntaxKind::{self, *},
+    SyntaxNode, SyntaxToken, TextRange, TextUnit, WalkEvent, T,
+};
+
+use std::collections::{HashMap, HashSet};
+
+// TODO make more like intellij's fmt model
+// Model holds immutable tree and mutable intermediate model to produce diff
+// the model will probably have to create its own tree to add the extra
+// info to each token/node:
+//
+// [1,2,3];
+// can be Brace token, ident, comma all of which knows their own rules and apply
+// them accordingly to produce [1, 2, 3]; ???
+
+#[derive(Debug)]
+// this will probably be an enum that holds a struct for space/indent/ect
+struct AtomEdit {
+    space_loc: SpaceLoc,
+    val: SpaceValue,
+    start: TextUnit,
+}
+
+impl AtomEdit {}
+
+#[derive(Debug)]
+pub(crate) struct SynBlock {
+    //indent: some enum?
+    kind: SyntaxKind,
+    node: SyntaxNode,
+    tokens: Vec<SyntaxToken>,
+    range: TextRange,
+    edits: Vec<AtomEdit>,
```

Comment:
I don't think that edits belong here. I *think* this should rather contain a mutable whitespace representation. We should get text edits only as a last step, as a result of diffing the model and the original tree.

See also the new comment at the original issue: I think each block should only remember its *preceding* whitespace.

Reply:
I agree pushing that of the the diffing model would make more sense.